src/netops.c

```diff
@@ -206,83 +206,100 @@ void gitno_connection_data_free_ptrs(gitno_connection_data *d)
+		error = GIT_EINVALIDSPEC;
+		goto done;
 	}
 
-	_host = url+u.field_data[UF_HOST].off;
-	_port = url+u.field_data[UF_PORT].off;
-	_path = url+u.field_data[UF_PATH].off;
-	_userinfo = url+u.field_data[UF_USERINFO].off;
+	has_host = !!(u.field_set & (1 << UF_HOST));
+	url_host = url + u.field_data[UF_HOST].off;
+	url_host_len = u.field_data[UF_HOST].len;
 
-	if (u.field_set & (1 << UF_HOST)) {
-		*host = git__substrdup(_host, u.field_data[UF_HOST].len);
-		GITERR_CHECK_ALLOC(*host);
-	}
+	has_port = !!(u.field_set & (1 << UF_PORT));
+	url_port = url + u.field_data[UF_PORT].off;
+	url_port_len = u.field_data[UF_PORT].len;
 
-	if (u.field_set & (1 << UF_PORT))
-		*port = git__substrdup(_port, u.field_data[UF_PORT].len);
-	else
-		*port = git__strdup(default_port);
-	GITERR_CHECK_ALLOC(*port);
+	has_path = !!(u.field_set & (1 << UF_PATH));
+	url_path = url + u.field_data[UF_PATH].off;
+	url_path_len = u.field_data[UF_PATH].len;
 
-	if (path) {
-		if (u.field_set & (1 << UF_PATH)) {
-			*path = git__substrdup(_path, u.field_data[UF_PATH].len);
-			GITERR_CHECK_ALLOC(*path);
-		} else {
-			git__free(*port);
-			*port = NULL;
-			git__free(*host);
-			*host = NULL;
-			giterr_set(GITERR_NET, "invalid url, missing path");
-			return GIT_EINVALIDSPEC;
-		}
+	has_userinfo = !!(u.field_set & (1 << UF_USERINFO));
+	url_userinfo = url + u.field_data[UF_USERINFO].off;
+	url_userinfo_len = u.field_data[UF_USERINFO].len;
+
+	if (has_host)
+		git_buf_decode_percent(&host, url_host, url_host_len);
+
+	if (has_port)
+		git_buf_put(&port, url_port, url_port_len);
+	else
+		git_buf_puts(&port, default_port);
+
+	if (has_path && path_out)
+		git_buf_decode_percent(&path, url_path, url_path_len);
+	else if (path_out) {
+		giterr_set(GITERR_NET, "invalid url, missing path");
+		error = GIT_EINVALIDSPEC;
+		goto done;
 	}
 
-	if (u.field_set & (1 << UF_USERINFO)) {
-		const char *colon = memchr(_userinfo, ':', u.field_data[UF_USERINFO].len);
+	if (has_userinfo) {
+		const char *colon = memchr(url_userinfo, ':', url_userinfo_len);
+
 		if (colon) {
-			*username = unescape(git__substrdup(_userinfo, colon - _userinfo));
-			*password = unescape(git__substrdup(colon+1, u.field_data[UF_USERINFO].len - (colon+1-_userinfo)));
-			GITERR_CHECK_ALLOC(*password);
+			const char *url_username = url_userinfo;
+			size_t url_username_len = colon - url_userinfo;
+
+			const char *url_password = colon + 1;
+			size_t url_password_len = url_userinfo_len - (url_username_len + 1);
+
+			git_buf_decode_percent(&username, url_username, url_username_len);
+			git_buf_decode_percent(&password, url_password, url_password_len);
 		} else {
-			*username = git__substrdup(_userinfo, u.field_data[UF_USERINFO].len);
+			git_buf_decode_percent(&username, url_userinfo, url_userinfo_len);
 		}
-		GITERR_CHECK_ALLOC(*username);
-
 	}
 
-	return 0;
+	if (git_buf_oom(&host) ||
+		git_buf_oom(&port) ||
+		git_buf_oom(&path) ||
+		git_buf_oom(&username) ||
+		git_buf_oom(&password))
```

Comment:
This leaks memory in case only some of these buffers OOM'd

Reply:
Yes.

Reply:
Okay

Reply:
Okay